### PR TITLE
Issue 414: Add support for headers

### PR DESCRIPTION
**Pull request by davidtrihy**

### CODE REVIEW

**Comment by Fr8Traindb** [comment]
Thanks David!

**Comment by VijetaH** [comment]
+1 for this feature!

**Comment by josephxsxn** [comment]
Having this would help our teams migrate away from some anti-patterns where we use the Kafka Key for header metadata today, it then results in sticky partitions which are consumed unevenly by downstream clients. Moving this metadata into a proper header would let the round robin function when missing a key to work as expected.

**Comment by R11** [comment]
+1 Currently having to work around this limitation is painful for our project

**Comment by davidtrihy** [comment]
@webmakersteve do you think you could take a look at this before the branch goes stale?

**Comment by iradul** [comment]
This works great. Thank you @davidtrihy!
However I think it would be much nicer if headers could look more [NodeJS like](https://nodejs.org/api/http.html#http_response_getheaders). For example instead of this:
```
const headers = [
    { "key1": "v1b" },
    { "key1": "v1a" },
    { "key2": "v2" },
    {
        "key3": "v3",
        "key4": "v4", // this pair will be ignored
    },
]
producer.produce('test_topic', null, Buffer.from('test message'), null, null, null, headers);
```
It would be much nicer to be able to do something like this:
```
const headers = {
    "key1": ["v1b", "v1a"],
    "key2": "v2",
    "key3": "v3",
}
producer.produce('test_topic', null, Buffer.from('test message'), null, null, null, headers);
```
This obvious also apples to the message consumption.

If you guys think this makes sense and you need help implementing this, please let me know.

**Comment by iradul** [comment]
Also worth noticing, this can't be built on Windows because `rdkafkacpp.h` doesn't include appropriate declarations for headers support. It works however with the latest _librdkafka_ `1.0.0` but some work still needs to be done to fix some other Windows build issues. I'll soon make PR to take care of those. That said I think this should be merged only after librdkafka dependecy gets upgraded to `1.0.0` or higher.

**Comment by davidtrihy** [comment]
@iradul thanks for the feedback, I do agree that the way you suggested is cleaner way of doing it, the only problem I have with it is that in C++ code it would be difficult to do this without using something like boost::variant or manipulating the memory in an unsavoury way.

Taking the producer as an example on line 453 of producer.cc when you get the value it's specified that it's a v8String and not an array and I can't immediately see how this would be done easily, however in saying that I'm not hugely familiar with nodejs/nan framework so if you have suggestions I'd be more than happy to hear them.

**Comment by iradul** [comment]
Look at the code that's parsing configuration object, I guess you could do something very similar to that (it's using `v8::Value` and `config_value->IsString()`):
https://github.com/Blizzard/node-rdkafka/blob/da9cbe3fc73f0f134d9cda364114b949e78b4706/src/common.cc#L509-L547

**Comment by josephxsxn** [comment]
Any word on this making its way in? We are going to start using this PR in a Fork which we would prefer not to...

File: examples/producer.md [hunk 1]
```diff
@@ -47,7 +47,10 @@ producer.on('ready', function(arg) {
     var key = "key-"+i;
     // if partition is set to -1, librdkafka will use the default partitioner
     var partition = -1;
-    producer.produce(topicName, partition, value, key);
+    var headers = [
+      { header: "header value" }
+    ]
+    producer.produce(topicName, partition, value, key, new Date(), "". headers);
```
Comment (cfinkenstadt) [inline examples/producer.md]: Why `"".` in front of the headers collection?  Looks like this should be `,` instead of `.`.

**Comment by jibugeorge** [comment]
+1. Any update on when this change will be available? We are looking into using this library to produce/consume kafka messages and we need  to be able to use header information.

**Comment by Fr8Traindb** [comment]
I would appreciate an update on this as well.

**Comment by edenhill** [comment]
Two small remarks;
 * Duplicate headers (keys) should be supported without overwriting previous entries.
 * Because of the possibility of duplicates, it is important to preserve the original header order.

**Comment by tiagoCMatias** [comment]
Any update on this?

**Comment by JaapRood** [comment]
> Two small remarks;
> 
> * Duplicate headers (keys) should be supported without overwriting previous entries.
> * Because of the possibility of duplicates, it is important to preserve the original header order.

@edenhill thanks for the info on that. It's probably a good idea to cover that in a test. It pretty much disallows the API of using object of headers straight up (`var headers = { a: '1', b: '' }`, because order is undefined and duplicate keys not allowed), which _is_ something a Node programmer might expect. Should prevent any future PR's from suggesting that without considering that requirement.

**Comment by jibugeorge** [comment]
@davidtrihy Any update on when this feature will be available?

File: examples/producer.md (hunk 1 above)
Comment (webmakersteve) [reply examples/producer.md]: Yeah I think this is a typo here. Would probably be good to get that fixed.

**Comment by webmakersteve** [comment]
Sorry everyone. I should probably post something else somewhere more public on the repo about this, but my time that I am able to have dedicated on this project has reduced substantially. I'm going to try to go through a lot of this today in an effort to catch up :(

**Comment by webmakersteve** [comment]
I'm going to merge this as is, with the README typo. I think the real solution for the object vs array of key value objects is going to be to encapsulate them in a top level type (`Headers`) which can be instantiated `fromObject` or `fromArray`. The produce call can then do checking and convert implicitly if needed.

**Pull request merged**